grant-project/zcash-grant-system
# Proposal staking

Closes #86.

### What this does
- adds `PROPOSAL_STAKING_AMOUNT=0.025` to backend environment variables
- adds `STAKING` proposal status
- adds `/<proposal_id>/stake` proposal end-point, similar to `contribute`
- adds `is_staked` field to proposal model (derived, no db field)
- adds `staking_contribution_confirmed` email
- submitting a DRAFT proposal will set it to STAKING
- will display `PaymentInfo` on `Final` step of create to solicit staking contribution
- when contribution confirms, it will email user & set proposal to PENDING
- if a proposal is REJECTED, it will ignore the STAKING status and skip to PENDING if the user decides to re-submit the Proposal
- ZCash -> Zcash in a few places

### What could be better
- testing is pretty minimal, it would be good to get a really solid set of Proposal lifecycle flow tests with All of the Combinations.


### Testing
1. set `backend` `.env` `PROPOSAL_STAKING_AMOUNT=0.025` (see .env.example)
1. create a proposal, submit the proposal, should show staking information
1. proposal should be STAKING under the User>Pending tab
1. the contribution should live under User>Funded
1. go ahead and fund on zcash, wait for confirmation
1. you should receive a Proposal staked email
1. proposal should now show as PENDING under User>Pending tab
1. use admin to REJECT the proposal
1. now re-submit the proposal through create flow, it should skip the staking step and automatically to to PENDING mode
1. use admin to APPROVE
1. you should now be able to PUBLISH the proposal

## Screens

#### Submit warning
![screen shot 2019-01-31 at 1 00 03 pm](https://user-images.githubusercontent.com/11430954/52078226-4e266680-2558-11e9-8298-337ffa8b9f3b.png)

#### Submit confirmation & staking contribution info
![screen shot 2019-01-31 at 1 00 26 pm](https://user-images.githubusercontent.com/11430954/52078267-65fdea80-2558-11e9-8bb9-0e377758363f.png)

#### Pending tab
![screen shot 2019-01-31 at 1 00 42 pm](https://user-images.githubusercontent.com/11430954/52078374-a8bfc280-2558-11e9-8503-14a38ee7b2ba.png)

#### Staking contribution confirmation email
![screen shot 2019-01-31 at 1 03 22 pm](https://user-images.githubusercontent.com/11430954/52078402-baa16580-2558-11e9-8157-ef8d7e80b551.png)
